hal/halx86/smp/i386/apentry.S

```diff
@@ -0,0 +1,136 @@
+ *              Copyright 2023 Victor Perevertkin <victor.perevertkin@reactos.org>
+ */
+
+#include <asm.inc>
+#include <ks386.inc>
+
+#define ZERO_OFFSET(f) (f - _HalpAPEntry16)
+#define PS(f) (f - _HalpAPEntryData)
+
+PUBLIC _HalpAPEntry16
+PUBLIC _HalpAPEntryData
+PUBLIC _HalpAPEntry32
+PUBLIC _HalpAPEntry16End
+
+.code16
+_HalpAPEntry16:
+    cli
+
+    /* Calculate the flat base address */
+    mov ebp, cs
+    shl ebp, 4
+
+    /* Use flat addressing */
+    xor eax, eax
+    mov ds, eax
+
+#ifdef _USE_ML
+    data32 lgdt fword ptr cs:[ZERO_OFFSET(Gdtr)]
+    data32 lidt fword ptr cs:[ZERO_OFFSET(Idtr)]
+#else
+    data32 lgdt cs:[ZERO_OFFSET(Gdtr)]
+    data32 lidt cs:[ZERO_OFFSET(Idtr)]
+#endif
+
+    /* Load temp page table */
+    mov eax, cs:[ZERO_OFFSET(PageTableRoot)]
+    mov cr3, eax
+
+    mov eax, cr0
+    or eax, HEX(80000001)
+    mov cr0, eax
+
+.align 4
+    /* Long jump, 32bit address */
+    .byte HEX(66)
+    .byte HEX(EA)
+_HalpAPEntryData:
+_APEntryJump32Offset:
+    .long 0
+_APEntryJump32Segment:
+    .long 8
+SelfPtr:
+    .long 0
+PageTableRoot:
+    .long 0
+ProcessorState:
+    .long 0
+Gdtr_Pad:
+    .short 0 // Pad
+Gdtr:
+    .short 0 // Limit
+    .long 0 // Base
+Idtr_Pad:
+    .short 0 // Pad
+Idtr:
+    .short 0 // Limit
+    .long 0 // Base
+_HalpAPEntry16End:
+.endcode16
+
+.code32
+_HalpAPEntry32:
+    /* Set the Ring 0 DS/ES/SS Segment */
+    mov ax, HEX(10)
+    mov ds, ax
+    mov es, ax
+    mov ss, ax
+    mov gs, ax
+
+    /* Load ProcessorState pointer */
+    mov esi, [ebp + ZERO_OFFSET(ProcessorState)]
+
+    mov eax, [esi + PsContextFrame + CsSegDs]
+    mov ds, eax
+    mov eax, [esi + PsContextFrame + CsSegEs]
+    mov es, eax
+    mov eax, [esi + PsContextFrame + CsSegSs]
+    mov ss, eax
+    mov eax, [esi + PsContextFrame + CsSegFs]
+    mov fs, eax
+    mov eax, [esi + PsContextFrame + CsSegGs]
+    mov gs, eax
+
+    /* Write CR registers with ProcessorState values */
+    mov eax, [esi + PsSpecialRegisters + SrCr3]
+    mov cr3, eax
+    mov eax, [esi + PsSpecialRegisters + SrCr4]
+    mov cr4, eax
+
+    /* Load debug registers */
+    mov eax, [esi + PsSpecialRegisters + SrKernelDr0]
+    mov dr0, eax
+    mov eax, [esi + PsSpecialRegisters + SrKernelDr1]
+    mov dr1, eax
+    mov eax, [esi + PsSpecialRegisters + SrKernelDr2]
+    mov dr2, eax
+    mov eax, [esi + PsSpecialRegisters + SrKernelDr3]
+    mov dr3, eax
+    mov eax, [esi + PsSpecialRegisters + SrKernelDr6]
+    mov dr6, eax
+    mov eax, [esi + PsSpecialRegisters + SrKernelDr7]
+    mov dr7, eax
+
+    /* Load TSS */
+    ltr word ptr [esi + PsSpecialRegisters + SrTr]
+
+    /* Load AP Stack */
+    mov esp, [esi + PsContextFrame + CsEsp]
+
+    /* Load flags */
+    mov eax, [esi + PsContextFrame + CsEflags]
+    sahf
+
+    /* Load LoaderBlock */
+    mov ecx, [esi + PsContextFrame + CsEax]
+    push ecx
+
+    /* Load Eip into ebx and call KiSystemStartup */
+    mov ebx, [esi + PsContextFrame + CsEip]
+    call ebx
```

Comment:
What's the advantage of doing this instead of doing what you did before: long ret to the kernel entry point,
https://github.com/reactos/reactos/commit/9f2e8c0517e18e80721cd4e34022048583eb4adc#diff-4248c7a67a0f88c70f73c3f96d6ce8b6742e328c97c63efdc9a4ff4bf88dbdc1R117